---
title: "Analysing Ribosome profiling data with DOTSeq: Differential ORF Translation"
author: 
  - name: "Chun Shen Lim*"
    affiliation: |
      1 Department of Biochemistry, Faculty of Biomedical Sciences, University of Otago, Dunedin, New Zealand.  
      2 Genetics Otago, University of Otago, Dunedin, New Zealand.  
      3 Maurice Wilkins Centre for Molecular Biodiscovery, Centres of Research Excellence, New Zealand.  
    email: "chunshen.lim@otago.ac.nz"
  - name: "Gabrielle S.W. Chieng"
    affiliation: |
      1 Department of Biochemistry, Faculty of Biomedical Sciences, University of Otago, Dunedin, New Zealand.  
      2 Genetics Otago, University of Otago, Dunedin, New Zealand.
date: "2025"
output: 
  html_document:
    toc: true
    toc_float: true
    theme: united
    highlight: tango
vignette: >
  %\VignetteIndexEntry{DOTSeq}  
  %\VignetteEngine{knitr::rmarkdown}  
  %\VignetteEncoding{UTF-8}
package: DOTSeq
bibliography: DOTSeq.bib
---

```{r opts-chunk, echo=FALSE, include=FALSE}
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      eval = TRUE,
                      message = FALSE,
                      error = FALSE,
                      warning = FALSE)
```

# Introduction
DOTSeq is a statistical framework for identifying differentially translated open reading frames (ORFs) from Ribo-seq and matched RNA-seq datasets. Unlike traditional gene-level approaches, DOTSeq performs ORF-level analysis, enabling detection of:

- Differential ORF Usage (DOU): Changes in the relative usage of ORFs within the same gene across conditions.
- Differential Translation Efficiency (DTE): Changes in ribosome occupancy relative to RNA abundance across conditions.

DOTSeq uses a beta-binomial GLM implemented via `glmmTMB`, modelling translation-specific effects through an interaction term (`condition:strategy`). It supports:

- Flexible modeling of dispersion and random effects
- Post hoc contrasts via emmeans
- Adaptive shrinkage via ashr

# Load Library
First, we will load libraries required to run this vignette.

```{r setup, warning=FALSE}
library(DOTSeq)
library(SummarizedExperiment)
```

# Example Dataset
We use data from [Ly et al. (2024)](https://doi.org/10.1038/s41586-024-08088-3), who profiled translation initiation and elongation in HeLa cells synchronised at Mitotic Cycling, Mitotic Arrest, and Interphase. Two biological replicates were performed for each cell cycle stage. The raw sequencing reads are deposited on the NCBI Gene Expression Omnibus (GEO) under the accession number [GSE230189](http://www.ncbi.nlm.nih.gov/projects/geo/query/acc.cgi?acc=GSE230189).

```{r dir}
dir <- system.file("extdata", package = "DOTSeq")
list.files(dir)
```

# Input data
Both Ribo-seq and match RNA-seq reads have been preprocessed by adapter trimming using Cutadapt and aligning to the reference genome using STAR. Mapped reads were summarised at the ORF level based on a flattened GTF annotation using featureCounts as described in [README](https://github.com/compgenom/DOTSeq/blob/main/README.md).

DOTSeq requires a counts matrix from featureCounts, where the first 6 columns correspond to ORF level annotation and all columns after that correspond to the counts for Ribo- and RNA-seq samples. To speed up the vignette execution we will use a subset of input data. 

```{r read-in-count-file}
cnt <- read.table(file.path(dir, "featureCounts.cell_cycle_subset.txt.gz"), header=TRUE, comment.char ='#')
names(cnt) <- gsub(".*(SRR[0-9]+).*", "\\1", names(cnt))
```

We can check the first five lines of the counts matrix:
```{r count-table-example}
head(cnt)
```

# ORF annotation files 
DOTSeq requires a flatten GTF and BED files with discrete, non-overlapping ORF annotation describing the genomic locations of ORFs. These annotation files, were generated using the orf_to_gtf.py wrapper from the RIBOSS engine as described in [README](https://github.com/compgenom/DOTSeq/blob/main/README.md).

Here, we specify the file path for the annotation files:

```{r ref}
flat <- file.path(dir, "gencode.v47.orf_flattened_subset.gtf.gz")
bed <- file.path(dir, "gencode.v47.orf_flattened_subset.bed.gz")
```

# Prepare Condition Table
A condition (metadata) table is required to define the experimental design for both Ribo-seq and RNA-seq samples. 
This table must include four essential columns: run (sample identifier), strategy (sequencing type, either "ribo" or "rna", or "0" and "1"), replicate (biological replicate number), and condition (biological condition, such as treatment or control). 

Additional columns may also be included to capture study-specific factors, for example batch information or other variables that could influence the analysis.

```{r condition-table}
meta <- read.table(file.path(dir, "metadata.txt.gz"))
names(meta) <-  c("run","strategy","replicate","treatment","condition")
head(meta)
```
In this example, we will compare Mitotic Cycling and Interphase, and focus on translating ribosomes were stalled using cyclohexamide (chx) during Ribo-seq.

```{r extract-chx}
cond <- meta[meta$treatment=="chx",] # extract only samples processed using cyclohexamide 
cond$treatment <- NULL # remove the treatment column
```

# Running DOTSeq
The `DOTSeq()` wrapper performs:
- Construction of two SummarizedExperiment objects (`DOTSeqDataSet()`)
- model fitting via `glmmTMB` (`fitDOU()`) and `DESeq2`
- Post hoc contrasts via `emmmeans` (`testDOU()`)
- Shrinkage of effect size via `ashr` (`testDOU()`). 

`DOTSeqDataSet()`, `fitDOU()`, and `testDOU()` can be ran separately. For simplicity, here we will use the `DOTSeq()` wrapper.

```{r dotseq, warning=FALSE}
m <- DOTSeq(
  count_table = cnt, 
  condition_table = cond, 
  flattened_gtf = flat, 
  bed = bed)
```

In this case Interphase was automatically assigned as the baseline. The target and baseline conditions can explicitly defined as `target = "Mitotic_Cycling"` and `baseline = "Interphase"`.

By default, `DOTSeq()` uses the `formula  = ~ condition * strategy` and `dispersion_modeling = "auto"`. We can customise the conditional and dispersion formulas to include batch and random effects.

# Accessing results
DOTSeq returns:

- sumExp: DOU results (beta-binomial GLMs)
- dds: DTE results (negative binomial GLMs)

Since the DTE module is based on `DESeq2`, we use dds for the naming convention.

```{r dotseq-objects}
names(m)
```

We store the post hoc results in the metadata of slot, which can be returned by:

```{r posthoc}
dou_results <- metadata(m$sumExp)$interaction_results
dte_results <- metadata(m$dds)$interaction_results
```

The key results for DOU are log-odds fold-change (PosteriorMean) and Local False Sign Rate (lfsr). These values were computed by `ashr`'s adaptive shrinkage estimator using coefficients (betahat) and their standard errors (sebetahat) estimated by beta-binomial GLMs as priors.

```{r dou}
dou_results
```

The DTE results were obtained from `lfcShrink()` within the `DOTSeq()` wrapper. We use `type = ashr` to compute all pairwise contrasts.

```{r dte}
dte_results
```

We can also inspect the `rowData` and `rowRanges` of the SummarizedExperiment object with ORF level annotation.

```{r rowdata}
rowdata <- rowData(m$sumExp)
rowdata
```

```{r rowranges}
rowRanges(m$sumExp)
```

# Visualisation
For plotting, we will slice the data frames for Mitotic_Cycling vs Interphase.

```{r merge}
dou_mci <- dou_results[dou_results$contrast == "Mitotic_Cycling - Interphase", ]
dte_mci <- dte_results[dte_results$contrast == "Mitotic_Cycling - Interphase", ]

results <- merge(dou_mci, dte_mci, by = "orf_id", all = TRUE)
```

DOTSeq provides a flexible plotting function `plotDOT()` for visualizing DOU or DTE results. To avoid re-downloading gene symbols in future plots, we will fetch the gene symbols via `biomaRt` by specifying `plot_types = NULL`. However, this step is optional.
 
```{r id-mapping}
id_mapping <- plotDOT(results = results, rowdata = rowdata, plot_types = NULL)
```

Let's generate a Venn diagram.

```{r plot-venn}
plotDOT(results = results, plot_types = "venn")
```

We will generate composite scatter plots comparing DTE and DOU. The composite scatter plots include marginal distributions of DTE and DOU along each axis. Points are colour-coded by significance or ORF types.

```{r plot-composite}
plotDOT(results = results, rowdata = rowdata, plot_types = "composite")
```

Finally, let's generate volcano plots and a heatmap for visualising DOU results.

```{r plot-volcano}
plotDOT(results = results, rowdata = rowdata, id_mapping = id_mapping, plot_types = "volcano")
```

```{r plot-heatmap, fig.width=4, fig.height=8, warning=FALSE}
plotDOT(results = results, rowdata = rowdata, id_mapping = id_mapping, plot_types = "heatmap")
```

# Session Info
```{r sessionInfo}
sessionInfo()
```
